### ISSUE TITLE
Warning: React version not specified in eslint-plugin-react settings.

### ISSUE THREAD

**Issue by kud** [yannickcr/eslint-plugin-react#1955]
Hello,

Since I've updated this plugin, i've got a message like `Warning: React version not specified in eslint-plugin-react settings.`.

This is true, I didn't specify the version of react, and this happens due to https://github.com/yannickcr/eslint-plugin-react/pull/1857/commits/8738e5917709c0cba0e7dd45289356f8a1d77384#diff-4e51c5c91adee02670adaca34f7fd7fdR17

However, when I read the doc, I've got this:

`"version": "15.0", // React version, default to the latest React stable release`

It's said "by default, it uses the latest version of react".

So why displaying a warning if there's a default and this is what I want? as I keep my react often upgraded.

**Comment by ljharb** [comment]
(linking to your comment here: https://github.com/yannickcr/eslint-plugin-react/pull/1857#issuecomment-415091579)

**Comment by ljharb** [comment]
Just because there's a default doesn't mean you shouldn't always be providing an explicit value.

Repeating https://github.com/yannickcr/eslint-plugin-react/pull/1857#issuecomment-415171990:
> i could see adding a separate option for "detect" (using the "version" as a fallback), which did require('react').version - want to open a PR that adds that?
> (Note that it'd have to use `process.cwd()`, and the `resolve` package, to get the proper require path)

**Comment by kud** [comment]
Yes, it's true.

For the moment, I use this workaround: `"version": "999.999.999"`.

**Comment by chrisdothtml** [comment]
+1 for supporting `latest` as the value

**Comment by marcofugaro** [comment]
Yeah please I don't want to update this config each time a new react comes out

**Comment by ljharb** [comment]
You have to update your package.json every time that happens; I’m not sure also updating eslintrc is much harder.

**Comment by marcofugaro** [comment]
Yeaaah... I don't have it in a package.json, I have it in the eslint config shared for the team, and each time I would have to upgrade deps, update by hand, publish a new version of the tool. 

Most importantly, I upgrade my dependencies using `yarn upgrade-interactive --latest`, not by hand. I (or other mantainers) might forget to update this version by hand.

**Comment by chrisdothtml** [comment]
We also use a shared eslint config and this would make it quite annoying every upgrade. If "latest" isn't supported, everyone's just gonna use "999.999.999" instead, which could lead to issues in the future if this plugin assumes that property contains their actual React version

**Comment by kud** [comment]
https://github.com/yannickcr/eslint-plugin-react/issues/1955#issuecomment-416659624

Well, I use `ncu` to update my packages, and yes it's not hard to update `.eslintrc` but you *have to* remember this and it's not something I want. :D 

I'm using `999.999.999` for the moment.

**Comment by bhardy** [comment]
I am getting a similar issue where I receive this warning when I only add `react` to my `plugins` and not my `extends`. We only want to grab 2 rules from `react`. I have the settings object in my `eslintrc.json` with the `"settings": { "react": { "version" : "16.3.2" }}`, along with rules:
```
  "rules": {
    "semi": ["error", "always"],
    "eqeqeq": "off",
    "react/jsx-no-bind": "off",
    "react/no-deprecated": "error",
    "no-console": "error"
  }
```

I still receive the warning, if I put `react` (or `"plugin:react/recommended"` into `extends` I no longer receive the warning, but I don't want to extend the rules.

Is this the same issue, a different issue, or am I doing something wrong?

Thank you!

**Comment by ljharb** [comment]
@bhardy that seems like a different issue; if you have the settings object you shouldn't get a warning at all.

**Comment by bhardy** [comment]
@ljharb would you like me to create a new issue?

**Comment by ljharb** [comment]
@bhardy yes please :-)

**Comment by dylang** [comment]
I'm using `settings` and `eslint` isn't passing it to `eslint-plugin-react`.

This is all that's in the `context`:

```
{ id: 'react/no-unused-prop-types',
  options: [],
  report: [Function: report] }
```

This is what our settings looks like:

```
 // https://eslint.org/docs/user-guide/configuring#adding-shared-settings
    settings: {
        react: {
            // https://github.com/yannickcr/eslint-plugin-react/issues/1955
            version: '999.999.999'
        }
    }
```

**Comment by mohsinulhaq** [comment]
I tried `{"version": "latest"}`, and it's working fine. Why are we pushing for '999.999.999'?

**Comment by ljharb** [comment]
@mohsinulhaq it would be more explicit than relying on React never hitting v999

**Comment by mohsinulhaq** [comment]
but there is no downside of using `latest` with regards to linting, right?

**Comment by ljharb** [comment]
The downside would be a lack of explicit configuration with respect to which React version you’re using, i suppose.

**Comment by anilanar** [comment]
This is possibly a bug. There are editor plugins that rely on exit code and stdout of `prettier-eslint` for auto-formatting.

As a formatter, `prettier-eslint` **must not** log something to `stdout` **when exit code is 0**.

**Comment by bertho-zero** [comment]
@anilanar Even when the version is filled the warning appears with `prettier-eslint`, I fixed that and I wait for an answer. (https://github.com/prettier/prettier-eslint/pull/194)

**Comment by manuelbieh** [comment]
Hmm had (or let's say: have) the same issue. Since I'm using an `.eslintrc.js` I was able to use the following workaround which seems to work. At least I don't see this warning anymore:

```
settings: {
    react: {
        version: require('./package.json').dependencies.react,
    },
},
```

🤷‍♂️

**Comment by ljharb** [comment]
@manuelbieh `require('react/package.json').version`; but once the next release goes out, you can use "detect".

**Comment by manuelbieh** [comment]
Ah right, even better.

Detect, also cool! Thanks 👍

**Comment by ljharb** [comment]
[v7.12.0](https://github.com/yannickcr/eslint-plugin-react/releases/tag/v7.12.0) is released.

**Comment by mohsinulhaq** [comment]
Cool! Thanks

**Comment by jaydenseric** [comment]
That's ok, just publish a new default in a new major version.

**Comment by ljharb** [comment]
When we get around to the next major, we surely will. There’s no hurry tho; you can maually set it for yourself.

**Comment by aczekajski** [comment]
The `detect` option is missing in docs, it took mi way too long to find this issue and thus an info about the new option.

**Comment by ljharb** [comment]
Good catch; a PR to add it to the docs would be appreciated.

**Comment by eeriee** [comment]
Add 'detect' to the setting but still encounter the warning.

**Comment by ljharb** [comment]
@eerie if updating to the latest doesn’t fix it, please file a new issue with your config and your plugin version

**Comment by craigcosmo** [comment]
my eslintrc is like this, where should i put the react version?

```json
{
	"parser": "babel-eslint",
	"parserOptions": {
		"ecmaFeatures": {
			"jsx": true,
			"modules": true
		}
	},
	"env": {
        "browser": true,
        "node": true,
        "es6":true
    },
	"extends":[
		"plugin:import/errors",
		"plugin:import/warnings"
	],
	"plugins": [
		"react",
		"json",
		"import"
	],
	"globals": {
		"new": true,
		"localStorage": true,
		"google": true,
		"process": true,
		"__dirname":true,
		"describe":true,
		"require":true,
		"module":true,
		"it":true,
		"Promise":true,
		"console":true,
		"__DEV__":true
	}
}
```

**Comment by ljharb** [comment]
@craigcosmo in `settings`, as a top-level property. See the readme.

**Comment by xylix** [comment]
This isn't documented in the readme anymore?

**Comment by ljharb** [comment]
@xylix https://github.com/yannickcr/eslint-plugin-react/blob/master/README.md#configuration

**Comment by Robertosousa86** [comment]
hello, I had the same problem, I managed to solve with:
`npm install react-peer react react-dom --save`
from: [https://www.npmjs.com/package/react-peer](url)
I hope this is useful  =)

**Comment by ljharb** [comment]
Better to use `npx install-peerdeps`; you shouldn’t need a dep on react-peer.

**Comment by ittroller** [comment]
3/2020, i fixed it in my .eslintrc
"settings": {
   "react": {
      "pragma": "React",
      "version": "16.12.0"
    }
}
hope it help you

**Comment by h-h-m** [comment]
settings: {
    react: {
        "version": "latest",
    },
},

**Comment by Hongbo-Miao** [comment]
I have **eslint-plugin-react** 7.20.5 installed in my root folder.
My root folder does not have **react**, while, my `client` sub folder has **react** 16.13.1 installed.

I tried to
```js
module.exports = {
  // ...
  settings: {
    react: {
      version: 'latest',
    },
  },
};
```

at my root **.eslintrc.js**, but got error

> error  Fragments are only supported starting from React v16.2. Please disable the `react/jsx-fragments` rule in ESLint settings or upgrade your version of React  react/jsx-fragments

So I think this `latest` is not really latest.

However, the workaround way works perfect.

```js
module.exports = {
  // ...
  settings: {
    react: {
      version: '999.999.999',
    },
  },
};
```

Thanks!

**Comment by ljharb** [comment]
@Hongbo-Miao `latest` isn't a valid option. It must be a version number, or `'detect'`.

eslint-plugin-react should not be installed or used in a package.json that doesn't also have react in it.

**Comment by Hongbo-Miao** [comment]
Thanks @ljharb I agree need react in it. That case I need splits **.eslintrc.js** to three: one in root folder, and one in **client**, one in **server** folder. However, I found it is a little hard to manage them when splits them. Here is [my repo](https://github.com/Hongbo-Miao/hongbomiao.com).
BTW, I saw you left Airbnb earlier, good luck with your next journey! 🎉🎉🎉

**Comment by samhou1988** [comment]
```
.eslintrc.js
```
```
settings: {
    react: {
        version: 'detect',
    },
}
```

**Comment by vadimtantale** [comment]
Вариант ниже помог!

> ```
> .eslintrc.js
> ```
> 
> ```
> settings: {
>     react: {
>         version: 'detect',
>     },
> }
> ```

**Comment by tjx666** [comment]
Still occur this issue.  @ljharb

repo: https://github.com/tjx666/scripting-listener

**Comment by ljharb** [comment]
@tjx666 interesting; the `context` object is missing `settings` entirely for some reason. Can you file a new issue?

**Comment by ljharb** [comment]
@tjx666 ah, it's because you have `settings` defined twice in your JS config, and the second one is overwriting the first one. If you merge the two, it works fine.

**Comment by daryazata** [comment]
Simply need to write exact version of react you re using ..

**Comment by ljharb** [comment]
@daryazata it supports "detect", so you shouldn't have to do that.

**Comment by MidKnightXI** [comment]
For people using .yml instead of `.js` or `.json` you just need to add it like that to your `.eslintrc.yml`

```yml
...
settings:
  react:
    version: "detect"
```

**Comment by juliehchung** [comment]
Hi there, I seem to still be having this issue on v7.31.8, and would love some insight if I'm just setting things up incorrectly.
This is a brand new React/TS project being set up, so what you see is all I have currently.
The eslint library is being loaded in successfully, and I don't have any errors, but I am still getting this warning:
![Screen Shot 2022-09-22 at 1 26 35 PM](https://user-images.githubusercontent.com/54191326/191844642-8aaa4ef5-71d0-4d24-896d-7e8bd8187ab4.png)

The contents of my `.eslintrc.js`:
```
module.exports = {
  root: true,
  env: {
    node: true,
    browser: true
  },
  parser: '@typescript-eslint/parser',
  parserOptions: {
    ecmaVersion: 2022,
    sourceType: 'module',
    ecmaFeatures: {
      jsx: true
    },
    settings: {
      react: {
        version: 'detect'
      },
      'import/resolver': {
        node: {
          paths: ['src'],
          extensions: ['.js', '.jsx', '.ts', '.tsx']
        }
      }
    }
  },
  extends: [
    'eslint:recommended',
    'plugin:react/recommended',
    'plugin:import/recommended',
    'plugin:import/typescript',
    'plugin:@typescript-eslint/recommended',
    'eslint-config-prettier'
  ],
  rules: {
    'react/react-in-jsx-scope': 'off',
    semi: [2, 'always']
  }
};
```

The contents of my `package.json`:
```
{
  "name": "project",
  "private": true,
  "version": "0.0.0",
  "scripts": {
    "dev": "vite",
    "build": "tsc && vite build",
    "preview": "vite preview",
    "lint": "eslint . --ext .ts,.tsx"
  },
  "dependencies": {
    "react": "^18.2.0",
    "react-dom": "^18.2.0"
  },
  "devDependencies": {
    "@types/react": "^18.0.17",
    "@types/react-dom": "^18.0.6",
    "@typescript-eslint/eslint-plugin": "^5.38.0",
    "@typescript-eslint/parser": "^5.38.0",
    "@vitejs/plugin-react": "^2.1.0",
    "eslint": "^8.23.1",
    "eslint-config-prettier": "^8.5.0",
    "eslint-plugin-import": "^2.26.0",
    "eslint-plugin-react": "^7.31.8",
    "prettier": "^2.7.1",
    "typescript": "^4.6.4",
    "vite": "^3.1.0"
  }
}

```

Current file structure:
![Screen Shot 2022-09-22 at 1 25 27 PM](https://user-images.githubusercontent.com/54191326/191844257-466e101c-6c03-4273-b50c-7dee136f8cb9.png)

**Comment by ljharb** [comment]
@juliehchung try running eslint on the command line. if it works, then the problem is in your editor; if not, then please file a new issue, ideally with a repro repo.

**Comment by juliehchung** [comment]
> @juliehchung try running eslint on the command line. if it works, then the problem is in your editor; if not, then please file a new issue, ideally with a repro repo.

Do you mind elaborating what you mean by this? The same warning still pops up when running eslint via the cli, but it will still run without errors.

**Comment by ljharb** [comment]
I mean running `npx eslint .`, outside of vscode.

If the same warning pops up, then I'd love a new issue with a repro repo :-)

**Comment by juliehchung** [comment]
> I mean running `npx eslint .`, outside of vscode.
> 
> If the same warning pops up, then I'd love a new issue with a repro repo :-)

Done here https://github.com/jsx-eslint/eslint-plugin-react/issues/3438
